File: ortools/math_opt/python/expressions_test.py
#!/usr/bin/env python3
# Copyright 2010-2025 Google LLC
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from absl.testing import absltest
from ortools.math_opt.python import expressions
from ortools.math_opt.python import model
from ortools.math_opt.python import variables


def _type_check_linear_sum(x: variables.LinearSum) -> None:
    """Does nothing at runtime, forces the type checker to run on x."""
    del x  # Unused.


class FastSumTest(absltest.TestCase):

    def test_variables(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        y = mod.add_binary_variable()
        z = 4
        result = expressions.fast_sum([x, y, z])
        _type_check_linear_sum(result)
        self.assertIsInstance(result, variables.LinearSum)
        result_expr = variables.as_flat_linear_expression(result)
        self.assertEqual(result_expr.offset, 4.0)
        self.assertDictEqual(dict(result_expr.terms), {x: 1.0, y: 1.0})

    def test_numbers(self) -> None:
        result = expressions.fast_sum([2.0, 4.0])
        _type_check_linear_sum(result)
        self.assertIsInstance(result, variables.LinearSum)
        result_expr = variables.as_flat_linear_expression(result)
        self.assertEqual(result_expr.offset, 6.0)
        self.assertEmpty(result_expr.terms)

    def test_heterogeneous_linear(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        result = expressions.fast_sum([2.0, 3.0 * x])
        _type_check_linear_sum(result)
        self.assertIsInstance(result, variables.LinearSum)
        result_expr = variables.as_flat_linear_expression(result)
        self.assertEqual(result_expr.offset, 2.0)
        self.assertDictEqual(dict(result_expr.terms), {x: 3.0})

    def test_heterogeneous_quad(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        result = expressions.fast_sum([2.0, 3.0 * x * x, x])
        self.assertIsInstance(result, variables.QuadraticSum)
        result_expr = variables.as_flat_quadratic_expression(result)
        self.assertEqual(result_expr.offset, 2.0)
        self.assertDictEqual(dict(result_expr.linear_terms), {x: 1.0})
        self.assertDictEqual(
            dict(result_expr.quadratic_terms),
            {variables.QuadraticTermKey(x, x): 3.0},
        )

    def test_all_quad(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        result = expressions.fast_sum([3.0 * x * x, x * x])
        self.assertIsInstance(result, variables.QuadraticSum)
        result_expr = variables.as_flat_quadratic_expression(result)
        self.assertEqual(result_expr.offset, 0.0)
        self.assertEmpty(result_expr.linear_terms)
        self.assertDictEqual(
            dict(result_expr.quadratic_terms),
            {variables.QuadraticTermKey(x, x): 4.0},
        )


class EvaluateExpressionTest(absltest.TestCase):

    def test_scalar_expression(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        sol = {x: 1.0}
        expr = 4.0
        self.assertEqual(expressions.evaluate_expression(expr, sol), 4.0)

    def test_linear(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        y = mod.add_variable()
        sol = {x: 1.0, y: 4.0}
        expr = 3 * x + y + 2.0
        self.assertAlmostEqual(
            expressions.evaluate_expression(expr, sol), 9.0, delta=1.0e-10
        )

    def test_quadratic(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        y = mod.add_variable()
        sol = {x: 1.0, y: 4.0}
        expr = 3 * x * y + y * y + 2.0 * x + 2.0
        self.assertAlmostEqual(
            expressions.evaluate_expression(expr, sol), 32.0, delta=1.0e-10
        )


if __name__ == "__main__":
    absltest.main()

File: ortools/javatests/com/google/ortools/mathopt/ExpressionsTest.java
// Copyright 2010-2025 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.ortools.mathopt;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.google.common.collect.ImmutableList;
import org.junit.jupiter.api.Test;

public final class ExpressionsTest {
  @Test
  public void linExpr_noArgs_emptyExpr() {
    HashLinearExpression expression = Expressions.linExpr();
    assertThat(expression.offset()).isEqualTo(0.0);
    assertThat(expression.getTerms()).isEmpty();
  }

  @Test
  public void linExpr_fromNumber_isNumber() {
    HashLinearExpression expression = Expressions.linExpr(4.0);
    assertThat(expression.offset()).isEqualTo(4.0);
    assertThat(expression.getTerms()).isEmpty();
  }

  @Test
  public void linExpr_fromCopy_isCopy() {
    var model = new Model();
    Variable x = model.addVariable();
    HashLinearExpression init = new HashLinearExpression().add(4.0).add(2.0, x);
    HashLinearExpression copy = Expressions.linExpr(init);
    assertThat(copy).isNotSameInstanceAs(init);
    assertThat(copy.offset()).isEqualTo(4.0);
    assertThat(copy.getTerms()).containsExactly(x, 2.0);
  }

  @Test
  public void sum_onList_hasSum() {
    var model = new Model();
    Variable x = model.addVariable();
    var summand =
        ImmutableList.of(x, Expressions.linExpr(2.0).add(4.0, x), Expressions.linExpr(-3.0));
    HashLinearExpression sum = Expressions.sum(summand);
    assertThat(sum.offset()).isEqualTo(-1.0);
    assertThat(sum.getTerms()).containsExactly(x, 5.0);
  }

  @Test
  public void sum_onEmptyList_isZero() {
    HashLinearExpression sum = Expressions.sum(ImmutableList.of());
    assertThat(sum.offset()).isEqualTo(0.0);
    assertThat(sum.getTerms()).isEmpty();
  }

  @Test
  public void sumStream_onList_hasSum() {
    var model = new Model();
    Variable x = model.addVariable();
    var summand =
        ImmutableList.of(x, Expressions.linExpr(2.0).add(4.0, x), Expressions.linExpr(-3.0));
    HashLinearExpression sum = summand.stream().collect(Expressions.sumCollector());
    assertThat(sum.offset()).isEqualTo(-1.0);
    assertThat(sum.getTerms()).containsExactly(x, 5.0);
  }

  @Test
  public void innerProduct_parallelLists_hasInnerProduct() {
    var model = new Model();
    Variable x = model.addVariable();
    // [3, 2] * [x, (3x - 1)] = -3x - 2
    HashLinearExpression innerProduct = Expressions.innerProduct(
        ImmutableList.of(3.0, 2.0), ImmutableList.of(x, Expressions.linExpr(1.0).subtract(3.0, x)));
    assertThat(innerProduct.offset()).isEqualTo(2.0);
    assertThat(innerProduct.getTerms()).containsExactly(x, -3.0);
  }

  @Test
  public void innerProduct_emptyLists_zero() {
    HashLinearExpression innerProduct =
        Expressions.innerProduct(ImmutableList.of(), ImmutableList.of());
    assertThat(innerProduct.offset()).isEqualTo(0.0);
    assertThat(innerProduct.getTerms()).isEmpty();
  }

  @Test
  public void innerProduct_unevenLists_throws() {
    assertThrows(IllegalArgumentException.class,
        () -> Expressions.innerProduct(ImmutableList.of(3.0), ImmutableList.of()));
  }

  @Test
  public void product_times2_isDouble() {
    var model = new Model();
    Variable x = model.addVariable();
    Variable y = model.addVariable();
    HashLinearExpression e = Expressions.linExpr(3.0).add(4.0, x).add(5.0, y);

    HashLinearExpression twoE = Expressions.product(2.0, e);

    assertThat(twoE.offset()).isEqualTo(6.0);
    assertThat(twoE.getTerms()).containsExactly(x, 8.0, y, 10.0);
    // Also check that the input was not modified.
    assertThat(e.offset()).isEqualTo(3.0);
    assertThat(e.getTerms()).containsExactly(x, 4.0, y, 5.0);
  }

  @Test
  public void quotient_div2_isHalf() {
    var model = new Model();
    Variable x = model.addVariable();
    Variable y = model.addVariable();
    HashLinearExpression e = Expressions.linExpr(6.0).add(8.0, x).add(10.0, y);

    HashLinearExpression halfE = Expressions.quotient(e, 2.0);

    assertThat(halfE.offset()).isEqualTo(3.0);
    assertThat(halfE.getTerms()).containsExactly(x, 4.0, y, 5.0);
    // Also check that the input was not modified.
    assertThat(e.offset()).isEqualTo(6.0);
    assertThat(e.getTerms()).containsExactly(x, 8.0, y, 10.0);
  }
}
